README.md
# node-red-contrib music

![node-red music logo](icons/node-red-music-logo.svg)
Designed for live-coding, this package offers a range of nodes to generate beats, subdivide them into bars and sections, and to generate sound synthesis instructions for the SuperCollider audio synthesis platform. Messages are sent from node-red to Supercollider via OSC.

# Installation

* You will need node-red: instructions are at <https://nodered.org/docs/getting-started/>
* Download and install [SuperCollider](https://supercollider.github.io/) (See note below on RPi)

# Start the engines

* Start SuperCollider and go to Language | Boot Server. (See note below on RPi)
* Start node-red, e.g. with the `node-red` command on the command-line
* Point your browser to <http://127.0.0.1:1880>
* In the pull-down menu (top right), choose "Manage Palette", go to the "Install" tab then search for "music" and install all of the nodes associated with node-red-contrib-music
* You also need to install the package node-red-contrib-osc as this is used to send the synthesis instructions via OSC to SuperCollider

# Create

There are some step-by-step instructions, including getting going on a Raspberry Pi at <https://stevenaeola.github.io/node-red-contrib-music/>.  If you're not on a raspberry Pi you can skip the first two sections. Then there's a section with a brief introduction to node-red if you don't already know. The specific details on node-red-contrib-music are at <https://stevenaeola.github.io/node-red-contrib-music/making-music> and onwards.

For video demonstration(s) see <https://www.youtube.com/playlist?list=PL5Tj84YnQVdIMXqM-mX85yGdWqBepwAIP>

# Raspberry Pi Installation

* Although node-red is bundled with Raspbian, it's not a very recent version, so you probably want to follow the instructions at <https://nodered.org/docs/hardware/raspberrypi>
* If you are working on a Raspberry Pi then scsynth (the part of SuperCollider that we need) is already installed with [Sonic Pi](https://sonic-pi.net/) and so is bundled with Raspbian. You will probably want to use a separate sound card with a RPi, a cheap USB one is fine, or you can use an expensive HiFi one if you wish.

* To start scsynth (part of SuperCollider) you need to run a shell script via the command line. Step-by-step instructions are at <https://stevenaeola.github.io/node-red-contrib-music/getting-started>. The shell script is included in the home directory of the `node-red-contrib-music` (versions for USB and non-USB audio). You may wish to make a symbolic link to run it from your home directory

```
ln -s ~/.node-red/node_modules/node-red-contrib-music/startsc-rpi-usb.sh ~/startsc.sh 
```

* You can use an [IQAudio sound card](http://www.iqaudio.com/downloads/IQaudIO.pdf) but you will need to edit your `/boot/config.txt` to comment out `#dtparam=audio=on` and add

```
dtoverlay=iqaudio-dacplus
dtoverlay=i2s-mmap
```


# Compatibility

node-red-contrib-music has been tested with

* OSX 10.13.4 on a MacBook Pro
* Windows 10
* PiTop OS release 2018-05-25 on a [pi-top](https://pi-top.com/) with RPi3 and a bog standard USB sound card
* A standalone RPi3 with an [IQAudio sound card](http://www.iqaudio.com/downloads/IQaudIO.pdf)
* SuperCollider 3.8.0 - 3.10.2
* node-red 0.18.3 - 0.20.3
* node v8.11.3 - v11.11.0
* npm 5.6.0 - 6.9.0

If you get it working on some other combination then do let us know

# Why node-red-music (NRM)?

NRM has the following aims:

* To be accessible. It has been tried it out with success in quite a few UK schools, both secondary (11-) and primary (-11).
* To be transferable, both in the tech and musical domains. Learning
  from NRM should be able to translate out of its own context. It sits
  on top of node-red, which itself sits on top of JavaScript. The
  basic syntax is graphical, but digging deeper the syntax relies on
  JSON. Snippets of pure JavaScript can be used, but are not
  necessary. From the musical perspective it tries to use standard
  musical notation e.g. scales start at 1, not 0. There are excellent
  domain-specific languages out there e.g. tidal, sonic-pi, but we
  emphasise transferability above brevity.
* To encourage collaboration. It is easy to synchronise across
  different instances. Many people see tech as an isolating and
  uncreative domain, we hope to address that misconception.
* To inter-work with other musical systems. Live sampling and replay is
  supported, it is also to easy to interact with other
  programs/systems via OSC, MIDI, WebSockets, UDP or whatever.
* To support alternative interfaces. Node-red is targeted at IoT
  applications. Using this support it is easy to control music
  via virtual interfaces such as twitter, or other web
  services. Physical devices can be controlled by or used to control
  NRM code, and many gadgets have contributed support via
  dedicated node types.
* To be open source. It builds on the work of others that have made their work freely available, and the node-red community is constantly building new ways of interacting. We often overlook the positive societal benefits of the open source movement, and model it offers for human progress through sharing and collaboration.


CHANGELOG.md
# Changelog

## v2.1.1 (22/02/2021)

#### enhancement

- [**enhancement**] examples all updated to v2 [#98](https://github.com/stevenaeola/node-red-contrib-music/issues/98)

---

## v1.0.0 (27/12/2020)
*No changelog for this release.*

---

## v1.1.0 (27/12/2020)
*No changelog for this release.*

---

## v2.0.0 (27/12/2020)
*No changelog for this release.*

---

## v2.0.1 (27/12/2020)
*No changelog for this release.*

---

## v2.1.0 (27/12/2020)

#### bug

- [**bug**] sonic-pi synthdefs should be sent by supercollider [#101](https://github.com/stevenaeola/node-red-contrib-music/issues/101)
- [**bug**] tabla_ghe1 sample should be mono [#93](https://github.com/stevenaeola/node-red-contrib-music/issues/93)
- [**bug**] setting node should clear value when closed [#92](https://github.com/stevenaeola/node-red-contrib-music/issues/92)
- [**bug**] volume setting with flow context should work [#90](https://github.com/stevenaeola/node-red-contrib-music/issues/90)
- [**bug**] elec_blip and elec_blip2 should be mono [#88](https://github.com/stevenaeola/node-red-contrib-music/issues/88)
- [**bug**] Volume control in example is not global [#87](https://github.com/stevenaeola/node-red-contrib-music/issues/87)
- [**bug**] misc_crow synth should be mono [#86](https://github.com/stevenaeola/node-red-contrib-music/issues/86)
- [**bug**] bd_fat synth should be mono [#79](https://github.com/stevenaeola/node-red-contrib-music/issues/79)

#### closed

- [**closed**] Setting scale to 'whole tone' in global setting doesn't work [#94](https://github.com/stevenaeola/node-red-contrib-music/issues/94)
- [**closed**] Reconnection of synth requires redeployment [#82](https://github.com/stevenaeola/node-red-contrib-music/issues/82)

#### enhancement

- [**enhancement**] Beat generator to have option to auto-start/continue on deployment [#95](https://github.com/stevenaeola/node-red-contrib-music/issues/95)
- [**enhancement**] Allow single-stepping of beat generator [#81](https://github.com/stevenaeola/node-red-contrib-music/issues/81)
- [**enhancement**] status for divide node [#80](https://github.com/stevenaeola/node-red-contrib-music/issues/80)
- [**enhancement**] Allow additions to sequencer values (e.g. notes, note lengths) with incoming messages [#63](https://github.com/stevenaeola/node-red-contrib-music/issues/63)
- [**enhancement**] Allow output of fx to another fx [#38](https://github.com/stevenaeola/node-red-contrib-music/issues/38)
- [**enhancement**] Make sequencer hold for indefinite period with "next" input [#6](https://github.com/stevenaeola/node-red-contrib-music/issues/6)

---

## v1.3.5 (01/09/2019)

#### bug

- [**bug**] Loading samples doesn't work on Windows [#77](https://github.com/stevenaeola/node-red-contrib-music/issues/77)

#### enhancement

- [**enhancement**] Add looper component for importing [#76](https://github.com/stevenaeola/node-red-contrib-music/issues/76)
- [**enhancement**] Add volume control to looper [#75](https://github.com/stevenaeola/node-red-contrib-music/issues/75)
- [**enhancement**] Use beat count to self-terminate looper synthdefs : no need to store ID [#52](https://github.com/stevenaeola/node-red-contrib-music/issues/52)

---

## v1.3.4 (17/06/2019)

#### bug

- [**bug**] Synth node does not list synthtypes [#74](https://github.com/stevenaeola/node-red-contrib-music/issues/74)

---

## v1.3.3 (26/05/2019)
*No changelog for this release.*

---

## v1.3.2 (08/05/2019)

#### enhancement

- [**enhancement**] Add icons for all node types [#72](https://github.com/stevenaeola/node-red-contrib-music/issues/72)

---

## v1.3.1 (26/04/2019)
*No changelog for this release.*

---

## v1.3.0 (24/04/2019)

#### bug

- [**bug**] Global bpm has no effect [#57](https://github.com/stevenaeola/node-red-contrib-music/issues/57)
- [**bug**] Match mono/stereo sample synths to sample types [#56](https://github.com/stevenaeola/node-red-contrib-music/issues/56)

#### closed

- [**closed**] Give synthcontrols as options in synth configuration [#59](https://github.com/stevenaeola/node-red-contrib-music/issues/59)

#### enhancement

- [**enhancement**] Add extra scale types [#71](https://github.com/stevenaeola/node-red-contrib-music/issues/71)
- [**enhancement**] Add documentation for soundfx node [#70](https://github.com/stevenaeola/node-red-contrib-music/issues/70)
- [**enhancement**] Add flanger soundfx [#68](https://github.com/stevenaeola/node-red-contrib-music/issues/68)
- [**enhancement**] Add distortion soundfx [#66](https://github.com/stevenaeola/node-red-contrib-music/issues/66)
- [**enhancement**] Add envelope to samples so they don't overlay [#65](https://github.com/stevenaeola/node-red-contrib-music/issues/65)
- [**enhancement**] Make all synths use sonic-pi format [#61](https://github.com/stevenaeola/node-red-contrib-music/issues/61)
- [**enhancement**] Reduce download size [#55](https://github.com/stevenaeola/node-red-contrib-music/issues/55)
- [**enhancement**] Categorise samples and allow search from selector [#48](https://github.com/stevenaeola/node-red-contrib-music/issues/48)
- [**enhancement**] Make build process for scsynths [#44](https://github.com/stevenaeola/node-red-contrib-music/issues/44)
- [**enhancement**] Use global volume for synths [#40](https://github.com/stevenaeola/node-red-contrib-music/issues/40)
- [**enhancement**] Use sonic-pi synths [#36](https://github.com/stevenaeola/node-red-contrib-music/issues/36)
- [**enhancement**] Give soundfx nodes initial fxcontrol values in configuration [#34](https://github.com/stevenaeola/node-red-contrib-music/issues/34)
- [**enhancement**] Label all nodes' input and outputs [#33](https://github.com/stevenaeola/node-red-contrib-music/issues/33)
- [**enhancement**] Make sid sample synth tuned [#26](https://github.com/stevenaeola/node-red-contrib-music/issues/26)

---

## v1.2.1 (31/03/2019)

#### bug


synth.js
const sc = require('./synth_common');
const multer = require('multer');
const glob = require('glob');
const path = require('path');
const _ = require('underscore');

const uploadDir = 'uploads';

// prefix for served GET requests
const synthtypesURL = 'node-red-contrib-music/synthtypes';
const userSampleURL = 'node-red-contrib-music/usersamples';

let synthtypes = require('./synthtypes');
for (let synthtype in synthtypes) {
    let synthcontrols = synthtypes[synthtype].synthcontrols || {};
    synthcontrols.pan = { 'description': 'position in stereo field: left (-1) or right (1)', 'min': -1, 'max': 1, 'default': 0 };
    synthtypes[synthtype].synthcontrols = synthcontrols;
}

module.exports = function (RED) {
    'use strict';

    RED.httpAdmin.get('/' + synthtypesURL, function (req, res) {
        // so we can edit syntypes.json without having to restart the server
        synthtypes = require('./synthtypes');
        res.json(synthtypes);
    });

    RED.httpAdmin.get('/' + userSampleURL, async function (req, res) {
        // list of already uploaded files
        let matches = [];
        matches.push('/' + uploadDir + '/*.wav');
        matches.push('/' + uploadDir + '/*.aiff');
        matches.push('/' + uploadDir + '/*.aif');

        let fpaths = [];
        for (let match of matches) {
            fpaths = fpaths.concat(glob.sync(match, { nocase: true, root: __dirname }));
        }
        let fnames = [];
        for (let fpath of fpaths) {
            fnames.push(path.basename(fpath));
        }
        res.json(fnames);
    });

// adapted from https://github.com/expressjs/multer/issues/439

    const storage = multer.diskStorage({
        destination (req, file, cb) {
          cb(null, uploadDir);
        },
        filename: function (req, file, cb) {
            cb(null, file.originalname);
          }
      });
    let upload = multer({ storage });

    RED.httpAdmin.post('/' + userSampleURL, upload.single('sample'), function (req, res) {
        // saving files that are uploaded
        const msg = 'Received file ' + req.file.path;
        res.send(msg);
    });

    function SynthNode (config) {
        const configurables =
        {
            root: { 'default': 'C4' },
            scale: { 'default': 'minor' },
            volume: { 'default': 50 },
            octave: { 'default': 0 },
            synthtype: { 'default': 'kick' },
            key: { 'default': 'C minor' },
            degree: { 'default': 'I' }
        };

        RED.nodes.createNode(this, config);
        const node = this;

        for (let conf in configurables) {
            configure(conf, config[conf]);
        }

        reset();

        this.on('input', function (msg) {
            if (msg.topic && msg.topic.startsWith('synthcontrol:')) {
                const synthcontrol = msg.topic.substring(13);
                const controlval = Number(msg.payload);
                setSynthcontrol(synthcontrol, controlval);
                return;
            }

            switch (msg.topic) {
                default:
                    switch (msg.payload) {
                        case 'tick':
                            configureTick(msg);
                            handleTickSynth(msg);
                            break;

                        case 'reset':
                            for (let conf in configurables) {
                                configure(conf, config[conf]);
                            }
                            reset();
                            // just this once the reset message is not propagated
                            break;

                        case 'stop':
                            // nothing to do
                            break;

                        case 'start':
                            // nothing to do
                            break;

                        default:
                            configureMsg(msg);
                    }
            }
        });

        this.on('close', function () {
            // nothing to do
        });

        function handleTickSynth (msg) {
            if (Array.isArray(msg.note)) {
                msg.note.forEach(function (noteVal) {
                    sendNote(noteVal, msg);
                });
            } else {
                sendNote(msg.note, msg);
            }
        }

        function sendNote (noteVal, msg) {
            let midi;
            if (isNaN(msg.midi)) {
                if (!isNaN(noteVal)) {
                    midi = note2midi(noteVal);
                }
            } else {
                midi = msg.midi;
            }

            const amp = sc.volume2amp(node);
            let details = { amp };

            if (midi) {
                Object.assign(details, { midi });
            }

            const bpm = msg.bpm || node.context().flow.get('bpm') || node.context().global.get('bpm');
            if (msg.beats) {
                details.beats = msg.beats;
                if (bpm) {
                    // set release to 0.25 beats
                    let release = 0.25 * 60 / bpm;
                    // tried subtracting release time but it seemed to leave gaps
                    let sustain = msg.beats * 60 / bpm;
                    Object.assign(details, { bpm, sustain, release });
                }
            }

            for (let p in node.parameters) {
                details[p] = node.parameters[p];
            }

            let synthtypeExtended = node.synthtype;
            if (node.synthtype === 'user-sample') {
                synthtypeExtended = synthtypeExtended + '#' + node.sampleName;
            }
            let noteMsg = {
                payload: 'tick',
                details,
                synthtype: synthtypeExtended
            };

            if (msg.timeTag) {
                noteMsg.timeTag = msg.timeTag;
            }

            // from v2.0 all synth instructions area sent this way
            node.send(noteMsg);
        }

        function setSynthcontrol (synthcontrol, value) {
            let synthcontrolSpecs = synthtypes[node.synthtype].synthcontrols || {};
            // all synths allow pan at least
            synthcontrolSpecs.pan = { 'note': { 'description': 'position in stereo field: left (-1) or right (1)', 'min': -1, 'max': 1, 'default': 0 } };
            if (!synthtypes[node.synthtype].synthcontrols[synthcontrol]) {
                node.warn('No such synthcontrol: ' + synthcontrol);
                return;
            }
            node.parameters[synthcontrol] = Number(value);
        }

        function reset () {
            node.tuned = config.tuned;
            node.sampleName = config.sampleName;
            node.parameters = node.parameters || {};
            if (config.synthcontrols) {
                for (let synthcontrol in config.synthcontrols) {
                    setSynthcontrol(synthcontrol, config.synthcontrols[synthcontrol]);
                }
            }
            setTimeout(() => {
                node.send({ topic: 'synthtype', payload: node.synthtype });
            }, 200);
        }

        function isTuned () {
            return synthtypes[node.synthtype] && synthtypes[node.synthtype].tuned;
        }

        // turn note number (degree of scale) into midi
        function note2midi (note) {
            if (Array.isArray(note)) {
                return _.map(note, note2midi);
            }

            if (!isTuned()) {
                return;
            }

            if (note === null) {
                return -1;
            }

            note = Number(note);

            const flow = node.context().flow;
            const global = node.context().global;

            var root = node.root;

            var scale = node.scale;

            var degree = node.degree;
            var globalkey = flow.get('key') || global.get('key');
            var globalroot;
            var globalscale;
            if (typeof globalkey === 'string') {
                var bits = globalkey.split(' ');
                globalroot = bits.shift();
                globalscale = bits.join(' ');
            }

            if (root === '') {
                root = flow.get('root') || global.get('root') || globalroot ||
                    configurables.root['default'];
            }

            if (scale === '') {
                scale = flow.get('scale') || global.get('scale') || globalscale ||
                    configurables.scale['default'];
            }

            scale = scale.toLowerCase();

            if (degree === '') {
                degree = flow.get('degree') || global.get('degree') || configurables.degree['default'];
            }

            // turn the degree into an offset

            const roman = {
                I: 1,
                II: 2,
                III: 3,
                IV: 4,
                V: 5,
                VI: 6,
                VII: 7,
                VIII: 8
            };

            var noteoffset = 0;
            if (_.isString(degree) && roman[degree.toUpperCase()]) {
                noteoffset = roman[degree.toUpperCase()] - 1;
            }

            var midiroot;

            if (isNaN(root)) {
                const name2midi = {
                    C: 60,
                    D: 62,
                    E: 64,
                    F: 65,
                    G: 67,
                    A: 69,
                    B: 71
                };

                var rootbits = root.toUpperCase().split('');
                var base = rootbits.shift();
                midiroot = name2midi[base];
                if (midiroot === undefined) {
                    node.warn('Scale root should be a midi number or start with a letter A-G');
                    return;
                }
                var next = rootbits.shift();
                if (next === '#' || next === 's') {
                    midiroot++;
                } else if (next === 'b') {
                    midiroot--;
                } else if (next !== undefined) {
                    rootbits.unshift(next);
                }
                if (rootbits.length > 0) {
                    var octave = Number(rootbits.join(''));
                    if (!isNaN(octave)) {
                        midiroot += (octave - 4) * 12;
                    }
                }
            } else {
                if (root < 0 && root > 127) {
                    node.warn('Scale root must be in range 0-127');
                    return;
                }
                midiroot = root;
            }

            note += noteoffset;

            const intervals = {
                minor: [0, 2, 3, 5, 7, 8, 10, 12],
                major: [0, 2, 4, 5, 7, 9, 11, 12],
                dorian: [0, 2, 3, 5, 7, 9, 10, 12],
                mixolydian: [0, 2, 4, 5, 7, 9, 10, 12],
                lydian: [0, 2, 4, 6, 7, 9, 11, 12],
                'major pentatonic': [0, 2, 4, 7, 9, 12],
                'minor pentatonic': [0, 3, 5, 7, 10, 12],
                blues: [0, 3, 5, 6, 7, 10, 12],
                chromatic: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
                'whole tone': [0, 2, 4, 6, 8, 10, 12],
                'octatonic': [0, 2, 3, 5, 6, 8, 9, 11, 12]
            };

            if (!intervals[scale]) {
                node.warn('Unrecognised scale: ' + scale);
                scale = configurables.scale['default'];
                node.warn('Using: ' + scale);
            }
            var offsets = intervals[scale];
            var midi = midiroot;

            // work out notes above the offset values by shifting up an octave
            while (note > offsets.length) {
                note -= offsets.length - 1;
                midi += offsets[offsets.length - 1];
            }

            // work out notes below the offset values by shifting down an octave
            // notes 0 and -1 are the same as +1

            if (note === 0 || note === -1) {
                note = 1;
            }

            var negative = false;
            while (note < 0) {
                negative = true;
                note += offsets.length - 1;
                midi -= offsets[offsets.length - 1];
            }

            midi += node.octave * 12;

            if (negative) {
                midi += offsets[note + 1];
            } else {
                midi += offsets[note - 1];
            }

            return midi;
        }

        function configureTick (msg) {
            for (var configurable in configurables) {
                var val = msg[configurable];
                if (val != null) {
                    configure(configurable, val);
                }
            }
        }

        function configureMsg (msg) {
            for (var configurable in configurables) {
                if (msg.topic === configurable) {
                    configure(configurable, msg.payload);
                }
            }
        }

        function configure (config, val) {
            if (!Object.keys(configurables).includes(config)) {
                node.warn(config + ' is not configurable');
                return;
            }

            var def = configurables[config].default;

            switch (config) {
                case 'volume':
                    var newVol = Number(val);
                    if (Number.isNaN(newVol)) {
                        if (Number.isNaN(node.volume)) {
                            node.volume = def;
                        }
                    } else {
                        node.volume = newVol;
                    }

                    node.volume = Math.min(100, Math.max(0, node.volume));

                    break;

                case 'key':
                    if (val) {
                        var bits = val.split(' ');
                        configure('root', bits.shift());
                        configure('scale', bits.shift());
                    }
                    break;

                case 'synthtype':
                    node[config] = val;
                    reset();
                    break;

                default:
                    node[config] = val;
            }
        }
    }

    RED.nodes.registerType('synth', SynthNode);
};
